src/pgen/cluster.cpp

```diff
@@ -244,195 +249,422 @@ void ProblemInitPackageData(ParameterInput *pin, parthenon::StateDescriptor *hyd
       parthenon::par_for(
-          DEFAULT_LOOP_PATTERN, "MagneticTower::AddInitialFieldToPotential",
-          parthenon::DevExecSpace(), a_kb.s, a_kb.e, a_jb.s, a_jb.e, a_ib.s, a_ib.e,
+          DEFAULT_LOOP_PATTERN, "cluster::ProblemGenerator::ApplyMagneticPotential",
+          parthenon::DevExecSpace(), kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
           KOKKOS_LAMBDA(const int &k, const int &j, const int &i) {
-            // Compute and apply potential
-            const Real x = coords.Xc<1>(i);
-            const Real y = coords.Xc<2>(j);
-            const Real z = coords.Xc<3>(k);
+            u(IB1, k, j, i) =
+                (A(2, k, j + 1, i) - A(2, k, j - 1, i)) / coords.Dxc<2>(j) / 2.0 -
+                (A(1, k + 1, j, i) - A(1, k - 1, j, i)) / coords.Dxc<3>(k) / 2.0;
+            u(IB2, k, j, i) =
+                (A(0, k + 1, j, i) - A(0, k - 1, j, i)) / coords.Dxc<3>(k) / 2.0 -
+                (A(2, k, j, i + 1) - A(2, k, j, i - 1)) / coords.Dxc<1>(i) / 2.0;
+            u(IB3, k, j, i) =
+                (A(1, k, j, i + 1) - A(1, k, j, i - 1)) / coords.Dxc<1>(i) / 2.0 -
+                (A(0, k, j + 1, i) - A(0, k, j - 1, i)) / coords.Dxc<2>(j) / 2.0;
+
+            u(IEN, k, j, i) += 0.5 * (SQR(u(IB1, k, j, i)) + SQR(u(IB2, k, j, i)) +
+                                      SQR(u(IB3, k, j, i)));
+          });
 
-            const Real r3 = pow(SQR(x) + SQR(y) + SQR(z), 3. / 2);
+      /************************************************************
+       * Add uniform magnetic field to the conserved variables
+       ************************************************************/
+      const auto &init_uniform_b_field = hydro_pkg->Param<bool>("init_uniform_b_field");
+      if (init_uniform_b_field) {
+        const Real bx = hydro_pkg->Param<Real>("uniform_b_field_bx");
+        const Real by = hydro_pkg->Param<Real>("uniform_b_field_by");
+        const Real bz = hydro_pkg->Param<Real>("uniform_b_field_bz");
+        parthenon::par_for(
+            DEFAULT_LOOP_PATTERN, "cluster::ProblemGenerator::ApplyUniformBField",
+            parthenon::DevExecSpace(), kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+            KOKKOS_LAMBDA(const int &k, const int &j, const int &i) {
+              const Real bx_i = u(IB1, k, j, i);
+              const Real by_i = u(IB2, k, j, i);
+              const Real bz_i = u(IB3, k, j, i);
+
+              u(IB1, k, j, i) += bx;
+              u(IB2, k, j, i) += by;
+              u(IB3, k, j, i) += bz;
+
+              // Old magnetic energy is b_i^2, new Magnetic energy should be 0.5*(b_i +
+              // b)^2, add b_i*b + 0.5b^2  to old energy to accomplish that
+              u(IEN, k, j, i) +=
+                  bx_i * bx + by_i * by + bz_i * bz + 0.5 * (SQR(bx) + SQR(by) + SQR(bz));
+            });
+        // end if(init_uniform_b_field)
+      }
+
+    } // END if(hydro_pkg->Param<Fluid>("fluid") == Fluid::glmmhd)
+  }
 
```

Comment:
Note that I didn't touch any of this. It's just marked because it was intended due to the explicit loop over blocks (at the ProblemGenerator is not the Mesh one and not the MeshBlock one)